src/main/java/com/epam/aidial/core/upstream/UpstreamRouteProvider.java

```diff
@@ -0,0 +1,111 @@
+        Map<String, TieredBalancer> newState = new HashMap<>();
+        Set<String> uniqueDeployments = new HashSet<>();
+
+        Map<String, Model> models = config.getModels();
+        updateDeployments(newState, oldState, uniqueDeployments, models.values());
+
+        Map<String, Application> applications = config.getApplications();
+        updateDeployments(newState, oldState, uniqueDeployments, applications.values());
+
+        Map<String, Addon> addons = config.getAddons();
+        updateDeployments(newState, oldState, uniqueDeployments, addons.values());
+
+        Map<String, Assistant> assistants = config.getAssistant().getAssistants();
+        updateDeployments(newState, oldState, uniqueDeployments, assistants.values());
+
+        LinkedHashMap<String, Route> routes = config.getRoutes();
+        updateRoutes(newState, oldState, uniqueDeployments, routes.values());
+
+        balancers = newState;
+    }
+
+    private static void updateRoutes(Map<String, TieredBalancer> newState, Map<String, TieredBalancer> oldState,
+                                     Set<String> uniqueDeployments, Collection<Route> routes) {
+        for (Route route : routes) {
+            String name = route.getName();
+
+            RouteEndpointProvider endpointProvider = new RouteEndpointProvider(route);
+            TieredBalancer previous = oldState.get(name);
+            updateDeployment(endpointProvider, previous, uniqueDeployments, newState);
+        }
+    }
+
+    private static void updateDeployments(Map<String, TieredBalancer> newState, Map<String, TieredBalancer> oldState,
+                                          Set<String> uniqueDeployments, Collection<? extends Deployment> deployments) {
+        for (Deployment deployment : deployments) {
+            String name = deployment.getName();
+
+            DeploymentUpstreamProvider endpointProvider = new DeploymentUpstreamProvider(deployment);
+            TieredBalancer previous = oldState.get(name);
+            updateDeployment(endpointProvider, previous, uniqueDeployments, newState);
+        }
+    }
+
+    private static void updateDeployment(UpstreamProvider upstream, TieredBalancer previous, Set<String> uniqueDeployments,
+                                         Map<String, TieredBalancer> newState) {
+        String name = upstream.getName();
+        if (!uniqueDeployments.add(name)) {
+            log.warn("Duplicate deployment name: {}", name);
+            return;
+        }
+        if (previous != null && isUpstreamsTheSame(upstream, previous)) {
+            newState.put(name, previous);
+        } else {
+            newState.put(name, new TieredBalancer(name, upstream.getUpstreams()));
+        }
+    }
+
+    private static boolean isUpstreamsTheSame(UpstreamProvider upstreamProvider, TieredBalancer balancer) {
+        return new HashSet<>(upstreamProvider.getUpstreams()).equals(new HashSet<>(balancer.getOriginalUpstreams()));
```

Comment:
upstreamProvider.getUpstreams().containsAll(new HashSet<>(balancer.getOriginalUpstreams()))?

Reply:
getOriginalUpstreams may be subset

Reply:
You create a new instance of HashSet. I suggest reuse 
```
(new HashSet<>(upstreamProvider.getUpstreams())).containsAll(balancer.getOriginalUpstreams())
```

Reply:
This is not a hot path of the code, so creating some garbage here - not so critical
You suggest to rewrite hashset equals with reusing 1 party, which means:
instead of simple calling
`new HashSet<>(upstreamProvider.getUpstreams()).equals(new HashSet<>(balancer.getOriginalUpstreams()));`
we should call
```
if (upstreamProvider.getUpstreams().size() != balancer.getOriginalUpstreams().size()) {
return false;
}
return upstreamProvider.getUpstreams().containsAll(new HashSet<>(balancer.getOriginalUpstreams()));
```

or merge those 2 into a single condition which is more complicated to read

1 prefer to have a simple single line condition here

Reply:
ok. let's do it.